system/known_hosts.py

```diff
@@ -185,54 +189,76 @@ def sanity_check(module,host,key,sshkeygen):
         module.fail_json(msg="Host parameter does not match hashed host field in supplied key")
 
 def search_for_host_key(module,host,key,path,sshkeygen):
-    '''search_for_host_key(module,host,key,path,sshkeygen) -> (current,replace)
+    '''search_for_host_key(module,host,key,path,sshkeygen) -> (found,replace_or_add,found_line)
 
-    Looks up host in the known_hosts file path; if it's there, looks to see
+    Looks up host and keytype in the known_hosts file path; if it's there, looks to see
     if one of those entries matches key. Returns:
-    current (Boolean): is host found in path?
-    replace (Boolean): is the key in path different to that supplied by user?
-    if current=False, then replace is always False.
+    found (Boolean): is host found in path?
+    replace_or_add (Boolean): is the key in path different to that supplied by user?
+    found_line (int or None): the line where a key of the same type was found
+    if found=False, then replace is always False.
     sshkeygen is the path to ssh-keygen, found earlier with get_bin_path
     '''
-    replace=False
     if os.path.exists(path)==False:
-        return False, False
+        return False, False, None
     #openssh >=6.4 has changed ssh-keygen behaviour such that it returns
     #1 if no host is found, whereas previously it returned 0
     rc,stdout,stderr=module.run_command([sshkeygen,'-F',host,'-f',path],
                                  check_rc=False)
     if stdout=='' and stderr=='' and (rc==0 or rc==1):
-        return False, False #host not found, no other errors
+        return False, False, None #host not found, no other errors
     if rc!=0: #something went wrong
         module.fail_json(msg="ssh-keygen failed (rc=%d,stdout='%s',stderr='%s')" % (rc,stdout,stderr))
 
-#If user supplied no key, we don't want to try and replace anything with it
+    #If user supplied no key, we don't want to try and replace anything with it
     if key is None:
-        return True, False
+        return True, False, None
 
     lines=stdout.split('\n')
-    k=key.strip() #trim trailing newline
-    #ssh-keygen returns only the host we ask about in the host field,
-    #even if the key entry has multiple hosts. Emulate this behaviour here,
-    #otherwise we get false negatives.
-    #Only necessary for unhashed entries.
-    if k[0] !='|':
-        k=k.split()
-        #The optional "marker" field, used for @cert-authority or @revoked
-        if k[0][0] == '@':
-            k[1]=host
-        else:
-            k[0]=host
-        k=' '.join(k)
+    new_key = normalize_known_hosts_key(key, host)
+
     for l in lines:
-        if l=='': 
-            continue
-        if l[0]=='#': #comment
+        if l=='':
             continue
-        if k==l: #found a match
-            return True, False #current, not-replace
-    #No match found, return current and replace
-    return True, True
+        elif l[0]=='#': #comment
```

Comment:
This is really picky, but the comment '#comment' here is now wrong - in the old version, we just skipped the comment line, but now we're parsing it for the line-number match.